sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py

```diff
@@ -570,18 +661,76 @@ def merge_results(last_result):
 
     return final_result
 
+  @staticmethod
+  def _build_watermark_updates(
+      runner_execution_context,  # type: execution.FnApiRunnerExecutionContext
+      stage_inputs,  # type: Iterable[str]
+      expected_timers,  # type: Iterable[translations.TimerFamilyId]
+      pcolls_with_da,  # type: Set[str]
+      transforms_w_splits,  # type: Set[str]
+      watermarks_by_transform_and_timer_family  # type: Dict[translations.TimerFamilyId, timestamp.Timestamp]
+  ) -> Dict[Union[str, translations.TimerFamilyId], timestamp.Timestamp]:
+    """Builds a dictionary of PCollection (or TimerFamilyId) to timestamp.
+
+    Args:
+      stage_inputs: represent the set of expected input PCollections for a stage
+      expected_timers: represent the set of TimerFamilyIds that the stage can
+        expect to receive as inputs.
+      pcolls_with_da: represent the set of stage input PCollections that had
+        delayed applications.
+      transforms_w_splits: represent the set of transforms in the stage that had
+        input splits.
+      watermarks_by_transform_and_timer_family: represent the set of watermark
+        holds to be added for each timer family.
+    """
+    updates = {
+    }  # type: Dict[Union[str, translations.TimerFamilyId], timestamp.Timestamp]
+
+    def get_pcoll_id(transform_id):
+      buffer_id = runner_execution_context.input_transform_to_buffer_id[
+          transform_id]
+      # For IMPULSE-reading transforms, we use the transform name as buffer id.
+      if buffer_id == translations.IMPULSE_BUFFER:
+        pcollection_id = transform_id
+      else:
+        _, pcollection_id = translations.split_buffer_id(buffer_id)
+      return pcollection_id
+
+    for pcoll in pcolls_with_da:
+      updates[pcoll] = timestamp.MIN_TIMESTAMP
+
+    for tr in transforms_w_splits:
+      pcoll_id = get_pcoll_id(tr)
+      updates[pcoll_id] = timestamp.MIN_TIMESTAMP
+
+    for timer_pcoll_id, ts in watermarks_by_transform_and_timer_family.items():
```

Comment:
It might be simpler to do something like

```
for timer_pcoll_id in expected_timers:
  updates[timer_pcoll_id] = watermarks_by_transform_and_timer_family.get(
      timestamp.MAX_TIMESTAMP)
```

than these two loops here. Or could timer_pcoll_id be in pcolls_with_da and/or transforms_w_splits?

Reply:
Done. Much better : )